### PR TITLE
feat(cli): agama profile autoyast command

### PR DESCRIPTION
**Pull request by jreidinger**
## Problem

Call `agama profile download` do for autoyast beside downloading also convert to json which is confusing to use e.g. `agama profile download ftp://test.com/autoyast.xml > profile.json`

## Solution

For downloading use `agama download` and for special autoyast processing regarding URL and also result use `agama profile autoyast`

### CODE REVIEW

File: rust/agama-cli/src/commands.rs [hunk 1]
```diff
@@ -72,4 +72,16 @@ pub enum Commands {
     /// not affect the root user.
     #[command(subcommand)]
     Auth(AuthCommands),
+
+    /// Download file from given URL
+    ///
+    /// Purpose of this command is to allow download from all supported schemas.
```
Comment (imobachgs) [inline rust/agama-cli/src/commands.rs:78]:
```suggestion
    /// The purpose of this command is to download files using AutoYaST supported schemas (e.g. device:// or relurl://).
```

File: rust/agama-cli/src/commands.rs [hunk 2]
```diff
@@ -72,4 +72,16 @@ pub enum Commands {
...
+    /// It can be used to download additional scripts, configuration files and so on.
+    /// For agama autoinstallation profiles it can be also used, but unless additional processing is required
```
Comment (imobachgs) [inline rust/agama-cli/src/commands.rs:80]:
```suggestion
    /// You can use it for downloading Agama autoinstallation profiles. However, unless you need additional processing,
```

File: rust/agama-cli/src/commands.rs [hunk 3]
```diff
@@ -72,4 +72,16 @@ pub enum Commands {
...
+    /// the "agama profile import" is recommended.
+    /// For autoyast conversion use directly "agama profile autoyast".
```
Comment (imobachgs) [inline rust/agama-cli/src/commands.rs:82]:
```suggestion
    /// If you want to convert an AutoYaST profile, use "agama profile autoyast".
```

Comment (imobachgs) [reply rust/agama-cli/src/commands.rs:82]: Additionally, we could put this URL to check the supported schemas: https://doc.opensuse.org/projects/autoyast/#Commandline-ay

File: rust/agama-cli/src/profile.rs [hunk 4]
```diff
@@ -9,32 +10,59 @@ use std::{
...
 use tempfile::TempDir;
+use url::Url;
 
 #[derive(Subcommand, Debug)]
 pub enum ProfileCommands {
-    /// Download the profile from a given location
-    Download { url: String },
+    /// Download the autoyast profile and print resulting json
+    Autoyast {
+        /// URL to autoyast XML, erb or rules&classes dir. Supports
```
Comment (imobachgs) [inline rust/agama-cli/src/profile.rs:19]:
```suggestion
        /// AutoYaST profile's URL. Any AutoYaST scheme, ERB and rules/classes are supported.
```

File: rust/agama-cli/src/profile.rs [hunk 5]
```diff
@@ -9,32 +10,59 @@ use std::{
...
+    /// Schema is available at /usr/share/agama-cli/profile.schema.json
+    Validate {
+        /// Local path to json file
+        path: String,
+    },
 
     /// Evaluate a profile, injecting the hardware information from D-Bus
-    Evaluate { path: String },
+    ///
+    /// For example of jsonnet profile see
```
Comment (imobachgs) [inline rust/agama-cli/src/profile.rs:34]:
```suggestion
    /// For an example of Jsonnet-based profile, see
```

File: rust/agama-cli/src/profile.rs [hunk 6]
```diff
@@ -9,32 +10,59 @@ use std::{
...
     /// Process autoinstallation profile and loads it into agama
     ///
     /// This is top level command that do all autoinstallation processing beside starting
     /// installation. Unless there is a need to inject additional commands between processing
     /// use this command instead of set of underlying commands.
-    /// Optional dir argument is location where profile is processed. Useful for debugging
-    /// if something goes wrong.
-    Import { url: String, dir: Option<PathBuf> },
+    Import {
+        /// URL where profile is located. Supported schemas are all that download supports and additionally
```
Comment (imobachgs) [inline rust/agama-cli/src/profile.rs:47]:
```suggestion
        /// Profile's URL. Supports the same schemas than te "download" command plus
```

File: rust/agama-cli/src/profile.rs [hunk 7]
```diff
@@ -9,32 +10,59 @@ use std::{
...
+        /// autoyast specific ones. Supported files are json, jsonnet, sh for agama profiles and erb, xml, directory
```
Comment (imobachgs) [inline rust/agama-cli/src/profile.rs:48]:
```suggestion
        /// AutoYaST specific ones. Supported files are json, jsonnet, sh for Agama profiles and ERB, XML, and rules/classes directories
```

File: rust/agama-cli/src/profile.rs [hunk 8]
```diff
@@ -9,32 +10,59 @@ use std::{
...
+        /// for autoyast support.
```
Comment (imobachgs) [inline rust/agama-cli/src/profile.rs:49]:
```suggestion
        /// for AutoYaST support.
```

File: rust/agama-cli/src/profile.rs [hunk 9]
```diff
@@ -9,32 +10,59 @@ use std::{
...
+        url: String,
+        /// Specific directory where all processing happens. If not specific temporary directory is used
```
Comment (imobachgs) [inline rust/agama-cli/src/profile.rs:51]:
```suggestion
        /// Specific directory where all processing happens. By default it uses a temporary directory
```

File: rust/agama-cli/src/profile.rs [hunk 10]
```diff
@@ -9,32 +10,59 @@ use std::{
...
-    let contents = reader.read()?;
-    out_fd.write_all(contents.as_bytes())?;
+pub fn download(url: &str, mut out_fd: impl Write) -> anyhow::Result<()> {
+    let mut handle = Easy::new();
+    handle.url(url)?;
+
+    let mut transfer = handle.transfer();
+    transfer.write_function(|buf|
+        // unwrap here is ok, as we want to kill download if we failed to write content
+        Ok(out_fd.write(buf).unwrap()))?;
```
Comment (imobachgs) [inline rust/agama-cli/src/profile.rs:63]: Do we get a proper error? Not tried, just asking...

File: rust/agama-lib/src/profile.rs [hunk 11]
```diff
@@ -1,49 +1,34 @@
...
-    pub fn new(url: &str) -> anyhow::Result<Self> {
-        let url = Url::parse(url)?;
-        Ok(Self { url })
+impl AutoyastProfile {
+    pub fn new(url: &Url) -> anyhow::Result<Self> {
+        Ok(Self { url: url.clone() })
     }
 
-    pub fn read(&self) -> anyhow::Result<String> {
+    pub fn read(&self, mut out_fd: impl Write) -> anyhow::Result<()> {
```
Comment (imobachgs) [inline rust/agama-lib/src/profile.rs:20]: np: I understand the reason, but looks weird that the function is called "read" and receives a "impl Write" argument :-)

File: rust/agama-lib/src/profile.rs [hunk 12]
```diff
@@ -1,49 +1,34 @@
...
         let path = self.url.path();
         if path.ends_with(".xml") || path.ends_with(".erb") || path.ends_with('/') {
-            self.read_from_autoyast()
+            let content = self.read_from_autoyast()?;
+            out_fd.write_all(content.as_bytes())?;
+            Ok(())
         } else {
-            self.read_from_url()
+            let msg = format!("Unsupported autoyast format at {}", self.url);
```
Comment (imobachgs) [inline rust/agama-lib/src/profile.rs:27]:
```suggestion
            let msg = format!("Unsupported AutoYaST format at {}", self.url);
```

**Review by imobachgs — COMMENTED** [review]
It looks good, but the documentation could be improved (added a few suggestions). Additionally, I guess we should update the changes file.

File: rust/agama-cli/src/profile.rs (hunk 10 above)
Comment (jreidinger) [reply rust/agama-cli/src/profile.rs:63]: in debug build for sure...regarding release I have to try it

File: rust/agama-lib/src/profile.rs (hunk 11 above)
Comment (jreidinger) [reply rust/agama-lib/src/profile.rs:20]: I can name it  `read_into`?

Comment (imobachgs) [reply rust/agama-lib/src/profile.rs:20]: I like it.

**Review by imobachgs — APPROVED** [review]

**Pull request merged**